### ISSUE TITLE
ErizoAgent - Launched new ErizoJS stucked in an infinite loop

### ISSUE THREAD

**Issue by rubenjimenez** [ging/licode#277]
Hi,

I'm trying Licode in an Ubuntu virtual machine through Vagrant + Virtual Box. When installation finished, everything seemed to work properly until I made:

```
./licode/scripts/initLicode.sh
```

The debugging messages started to show up in console but the next message was repeating itself in a infinite loop:

```
Running process
2015-03-24 16:26:10.115  - INFO: ErizoAgent - Launched new ErizoJS  303a8b26-d0d9-2945-4def-1f267ef9d05b
```

At last, It returns this error:

```
child_process.js:958
    throw errnoException(process._errno, 'spawn');
          ^
Error: spawn EMFILE
    at errnoException (child_process.js:1011:11)
    at ChildProcess.spawn (child_process.js:958:11)
    at exports.spawn (child_process.js:746:9)
    at launchErizoJS (/vagrant/licode/erizo_controller/erizoAgent/erizoAgent.js:100:24)
    at fillErizos (/vagrant/licode/erizo_controller/erizoAgent/erizoAgent.js:132:13)
    at ChildProcess.<anonymous> (/vagrant/licode/erizo_controller/erizoAgent/erizoAgent.js:112:9)
    at ChildProcess.emit (events.js:98:17)
    at maybeClose (child_process.js:766:16)
    at Process.ChildProcess._handle.onexit (child_process.js:833:5)
```

I installed Licode in an Ubuntu Amazon AWS instance as well and it worked as expected, but i'm suffering this glitch in local.

Any suggestion?

Thanks and great work!

**Comment by lodoyun** [comment]
Hi,
That is usually a problem with erizo compilation, usually because dependencies are missing.
What Ubuntu version are you running? 
Did you run installUbuntuDependencies or installLinuxDependencies first? 
If you did, can you run installErizo and paste the output?

**Comment by rubenjimenez** [comment]
Hi!

It must be what you said. I didn't recompile anything, but I tried your Vagrant deployment included in Licode repository and everything worked as expected. Right now i'm using it with no problem.

Although this is not related with the main topic, i got a little glitch testing your basic example. It didn't just connect to socket.io because the auto detected public IP was wrong. It failed when reaching this line:

```
room.connect();
```

I guessed that could be happening because of Vagrant so I modified the licode config file. I changed this line adding the localhost IP (with 'localhost' as value it didn't work either):

```
config.erizoController.publicIP = '127.0.0.1'; //default value: ''
```

With this change it worked.

The main problem is that I got the same result when I tried licode over an Amazon AWS instance. Even though the private IP was properly detected, the script was not capable to connect to socket.io. I'm guessing I'll have to change the previous line again for AWS in order to get it to work correctly.

Is this usual?

**Comment by lodoyun** [comment]
Yes, if you deploy parts of Licode behind a NAT (Not Recommended) you have to do a couple of things:
- For erizoController, edit the licode_config.js and set up `config.erizoController.publicIP`

- For erizoAgent you have two options:
 * Set up `config.erizoAgent.publicIP`: it should be pretty easy to automate this in AWS
 * Set up a stun server for erizo in `config.erizo.stunserver`, this will make all connections to Licode a little bit slower but should always work.

**Comment by rubenjimenez** [comment]
Hi Pedro,

I'm sorry for disturbing you so much, but i don't get Licode to work in AWS. In local finally everything got connected and it worked like a charm. But i have been trying Licode to work under AWS with no success.

I have tried a lot of different approaches with different kind of errors that at last i've got them fixed, but right now this is the error i'm getting and I'm stucked:

```
ubuntu@ip-10-72-136-241:~$ ./licode/scripts/initLicode.sh
801
2015-03-26 23:44:18.860  - INFO: RPC - Conected to rabbitMQ server
2015-03-26 23:44:18.867  - INFO: RPC - Exchange rpcExchange is open
2015-03-26 23:44:18.873  - INFO: RPC - Queue nuveQueue is open
2015-03-26 23:44:18.876  - INFO: RPC - ClientQueue amq.gen-wMkcNfb1E1Q9xWkq3gyXyf is open
[licode] Done, run basic_example/basicServer.js
ubuntu@ip-10-72-136-241:~$ Running process
2015-03-26 23:44:23.691  - INFO: ErizoAgent - Launched new ErizoJS  73f0a8fa-5c75-9450-9108-ed69b7f11c28
2015-03-26 23:44:23.816  - INFO: AMQPER - Exchange rpcExchange is open
2015-03-26 23:44:23.823  - INFO: AMQPER - Exchange broadcastExchange is open
2015-03-26 23:44:23.828  - INFO: AMQPER - ClientQueue amq.gen-AE6BbRc4NaD8rNLcwF6BJf is open
2015-03-26 23:44:23.842  - INFO: AMQPER - Queue ErizoAgent is open
2015-03-26 23:44:23.981  - INFO: AMQPER - Exchange rpcExchange is open
2015-03-26 23:44:23.987  - INFO: AMQPER - Exchange broadcastExchange is open
2015-03-26 23:44:23.990  - INFO: AMQPER - ClientQueue amq.gen-gtXcQIGdg4zHDNFwICDQQL is open
2015-03-26 23:44:24.078  - INFO: CloudHandler - private ip  10.72.136.241
2015-03-26 23:44:24.150  - INFO: CloudHandler - Error:  { [Error: getaddrinfo ENOTFOUND] code: 'ENOTFOUND', errno: 'ENOTFOUND', syscall: 'getaddrinfo' }
Error: getaddrinfo ENOTFOUND
    at errnoException (dns.js:37:11)
    at Object.onanswer [as oncomplete] (dns.js:124:16)

/home/ubuntu/licode/nuve/nuveAPI/cloudHandler.js:115
            callback('error');
            ^
TypeError: string is not a function
    at /home/ubuntu/licode/nuve/nuveAPI/cloudHandler.js:115:13
    at ClientRequest.<anonymous> (/home/ubuntu/licode/node_modules/aws-lib/lib/aws.js:62:14)
    at ClientRequest.emit (events.js:95:17)
    at CleartextStream.socketErrorListener (http.js:1548:9)
    at CleartextStream.emit (events.js:95:17)
    at Socket.onerror (tls.js:1456:17)
    at Socket.emit (events.js:117:20)
    at net.js:834:16
    at process._tickCallback (node.js:448:13)
2015-03-26 23:44:28.999  - INFO: ErizoController - CloudHandler does not respond
2015-03-26 23:44:37.006  - INFO: ErizoController - CloudHandler does not respond
2015-03-26 23:44:45.011  - INFO: ErizoController - CloudHandler does not respond
2015-03-26 23:44:53.019  - INFO: ErizoController - CloudHandler does not respond
2015-03-26 23:45:01.027  - INFO: ErizoController - CloudHandler does not respond
```

Currently i've got a Turn server configured and running, but i got the same error without it.

This is my licode config, all the parts i've modified:

```
config.cloudProvider = {};
config.cloudProvider.name = 'amazon';
//In Amazon Ec2 instances you can specify the zone host. By default is 'ec2.us-east-1a.amazonaws.com'
config.cloudProvider.host = 'ec2.eu-west-1a.amazonaws.com';
config.cloudProvider.accessKey = 'ACCESS-KEY';
config.cloudProvider.secretAccessKey = 'SECRET-KEY';

//Use undefined to run clients without Turn
config.erizoController.turnServer = {}; // default value: undefined
config.erizoController.turnServer.url = 'turn:54.75.251.18:3478?transport=tcp'; // default value: null
config.erizoController.turnServer.username = 'USER'; // default value: null
config.erizoController.turnServer.password = 'PASS'; // default value: null
```

All of this is running in Ubuntu 12.04 LTS, I have access to Turn server from http://54.75.251.18:3478, I have TCP ports 8080, 3000, 3001 and UDP ports 30000 - 31000 opened.

What am i missing? I've been looking for some insight in the Lynckia list and i found a guy who was getting the same error, but no one answered him.

Thanks!

**Comment by lodoyun** [comment]
the `cloudProvider.*`parts in the configuration can be used to create new erizoController instances on the fly in ec2.
You don't actually need to use them to use Licode in EC2. To quickly try this just install it normally and clear those values, you do need to set up the public IPs in the configuration. It should work exactly the same as in a local VM.

**Comment by rgwsk65** [comment]
I've got the same problem with ErizoAgent: it spams in loop
`ErizoAgent - Launched new ErizoJS xxxxxxxxxxxxx"`
until it stopped work.

For me the problem was method with which i made node listen at port 80.

When I gave capacities for node like that:
`sudo setcap cap_net_bind_service=+ep /usr/local/bin/node`
I had that issue.

I fixed it by removing this capacity, made my app listen at port 3065 (or any else) and redirect traffic from port 80 to 3065 via `iptables`.
